### PR TITLE
session: optimize the sharding algorithm for non-transactional DMLs

### PR DESCRIPTION
**Pull request by ekexium**
Signed-off-by: ekexium <ekexium@gmail.com>

### What problem does this PR solve?

Issue Number: ref #33485

Problem Summary:

Use an accurate sharding algorithm that makes the best effort to make the actual batch size equal to the batch size specified by the user. 
Using the original algorithm, when there are duplicated values, the worst case can be a super large batch containing all values. The PR avoids such cases.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by you06 — COMMENTED** [review]
rest LGTM

File: session/nontransactional.go [hunk 1]
```diff
@@ -304,29 +305,36 @@ func buildShardJobs(ctx context.Context, stmt *ast.NonTransactionalDeleteStmt, s
 			break
 		}
 
-		newStart := chk.GetRow(0).GetDatum(0, &rs.Fields()[0].Column.FieldType)
+		if len(jobs) > 0 && chk.NumRows()+currentSize < batchSize {
+			// not enough data for a batch
+			currentSize += chk.NumRows()
```
Comment (you06) [inline session/nontransactional.go:310]: Also need to update `currentEnd` in case it's the last chunk.

**Review by you06 — APPROVED** [review]

**Review by cfzjywxk — APPROVED** [review]

**Pull request merged**